### ISSUE TITLE
Streaming response not working

### ISSUE THREAD

**Issue by EVOsparkLP** [moe-mizrak/laravel-openrouter#13]
I have some issues to get the response as a stream. While the "normal" request works really fine, I get an error response, when I set 'stream' => true.

This is how I to use the streaming response. Maybe I am using the laravel-openrouter package in a wrong way, but as I could not find an example, I used some code that works with the openai-php/laravel package:

```
public function answerQuestion(Request $request)
{		
	$question = $request->data["question"];

	return response()->stream(function () use ($question) {

		$messageData = new MessageData([
			'content' => $question,
			'role'    => RoleType::USER,
		]);
		$chatData = new ChatData([
			'messages' => [
				$messageData
			],
			'model' => 'openai/gpt-4o',
			'temperature' => 0.2,
			'stream' => true,
		]);
		$stream = LaravelOpenRouter::chatRequest($chatData);	   
		
		foreach ($stream as $response) {
			$text = $response->choices[0]->delta->content;
			ServerEvent::send($text, "");
		}

	}, 200, [
		'Cache-Control' => 'no-cache',
		'Connection' => 'keep-alive',
		'X-Accel-Buffering' => 'no',
		'Content-Type' => 'text/event-stream',
		'Access-Control-Expose-Headers' => 'MessageID',
		'MessageID' => $message->id,
	]);
}
```

While calling the LaravelOpenRouter::chatRequest function, this error occurs:
Cannot assign null to property MoeMizrak\LaravelOpenrouter\DTO\ResponseData::$id of type string {"userId":"21154652-1287-44ae-ab78-5a066a29114a","exception":"[object] (TypeError(code: 0): Cannot assign null to property MoeMizrak\\LaravelOpenrouter\\DTO\\ResponseData::$id of type string at /var/www/html/vendor/spatie/data-transfer-object/src/Reflection/DataTransferObjectProperty.php:48)
[stacktrace]
#0 /var/www/html/vendor/spatie/data-transfer-object/src/Reflection/DataTransferObjectProperty.php(48): ReflectionProperty->setValue()
#1 /var/www/html/vendor/spatie/data-transfer-object/src/DataTransferObject.php(29): Spatie\\DataTransferObject\\Reflection\\DataTransferObjectProperty->setValue()
#2 /var/www/html/vendor/moe-mizrak/laravel-openrouter/src/OpenRouterRequest.php(116): Spatie\\DataTransferObject\\DataTransferObject->__construct()
#3 /var/www/html/vendor/moe-mizrak/laravel-openrouter/src/OpenRouterRequest.php(55): MoeMizrak\\LaravelOpenrouter\\OpenRouterRequest->formChatResponse()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(355): MoeMizrak\\LaravelOpenrouter\\OpenRouterRequest->chatRequest()

Maybe someone can verify if this is a real bug or provide an example how to use the streaming option in a proper way.
Thanks in advance!

**Comment by moe-mizrak** [comment]
@EVOsparkLP Thanks for flagging the issue. It is resolve now, you can install new version v1.0.3 which supports streaming.

Please check readme section [Stream Chat Request](https://github.com/moe-mizrak/laravel-openrouter?tab=readme-ov-file#stream-chat-request)
for how to use it.

Please don't hesitate to ask or flag any issue right away if you face any.

**Comment by EVOsparkLP** [comment]
@moe-mizrak Thanks for having a look at this issue. It took me some days to have a closer look at the streaming implementation.
Basically the streaming response seems to work, but to me it looks like some parts of the response get lost. While the rawData is always complete, the function filterStreamingResponse drops the chunks that where not fully included in one package.


This is the code I used for testing:
```
while(! $stream->eof()) {
    $rawResponse = $stream->read(1024);
    Log::info("RawResponse: ".$rawResponse);

    $response = LaravelOpenRouter::filterStreamingResponse($rawResponse);
    
    foreach ($response as $r) {
        $text = $r->choices[0]["delta"]["content"];
        Log::info("Filtered: ".json_encode($r));
    }
}
```

And this is how my log files look:
In this example the rawResponse contains 3 complete elements and one element that is incomplete:

```
data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":""},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// ""

data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":"Hallo"},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// "Hallo"

data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":"!"},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// "!"

data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":" Wie  
// " Wie // incomplete
```

The filtered response contains the first 3 elements (what is OK, because the next element is not complete, yet):
```
[2024-06-25 11:13:59] local.INFO: Filtered: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai\/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":""},"finish_reason":null}],"usage":null}
// ""

[2024-06-25 11:13:59] local.INFO: Filtered: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai\/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":"Hallo"},"finish_reason":null}],"usage":null}  
// "Hallo"

[2024-06-25 11:13:59] local.INFO: Filtered: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai\/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":"!"},"finish_reason":null}],"usage":null}  
// "!"
```

The next rawResponse continues with the incomplete message and sends some new elements:
```
[2024-06-25 11:13:59] local.INFO: RawResponse: "},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// continuing last incomplete message

data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":" geht"},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// " geht"

data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":"'s"},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// "'s"

data: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":" dir"},"finish_reason":null}],"system_fingerprint":"fp_3e7d703517"}
// " dir"
```

After this rawData is filtered, the filtered response does not contain the message that was incomplete in the first response. It contains only the new elements:
```
// Here I would expect to receive the "Wie"

[2024-06-25 11:13:59] local.INFO: Filtered: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai\/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":" geht"},"finish_reason":null}],"usage":null} 
// "geht"

[2024-06-25 11:13:59] local.INFO: Filtered: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai\/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":"'s"},"finish_reason":null}],"usage":null}  
// "'s"

[2024-06-25 11:13:59] local.INFO: Filtered: {"id":"gen-5r5R4klF00MFdDKxQ6B25PyZ6O4K","model":"openai\/gpt-4o","object":"chat.completion.chunk","created":1719306837,"choices":[{"index":0,"delta":{"role":"assistant","content":" dir"},"finish_reason":null}],"usage":null}  
// " dir"
```

Best regards!

**Comment by moe-mizrak** [comment]
@EVOsparkLP I see what is the issue, and it is obvious to me why data is getting lost. 
I fixed it in new version **v1.0.4** , please tyr it out. No data after filterStreamingResponse should be missing, it is buffering incomplete data and then give the ResponseData.
As you stated, raw response is already working fine.

I hope you will not see any issue now, thanks for contributing by flagging the issue such a clear way!

**Comment by EVOsparkLP** [comment]
I can confirm, that everything works as expected now. Thanks for fixing this issue!
